[WIP] Add dynamic scheduling operation mode

mtreinish opened the pull request:
This commit adds an opt-in scheduler option for dynamic scheduling.
Instead of partitioning the test list up-front based on historical
timing data this commit lets each worker ask for the next test
dynamically.  This is built using python's multiprocess module to
launch new workers instead of shelling out to call python via
subprocess.

This hopefully will provide a better worker balance since we will keep
each worker occupied until there are no more tests to be run. Instead
of trying to pack fill each work optimially up front. Additionally this
should hopefully improve the pdb story for users who use pdb with tests.
Since instead of spawning subprocesses calling python to invoke the
subunit runner and reading the subunit stream from stdout and instead
uses multiprocessing to fork workers and uses pipes to pass the subunit
streams between workers.

mtreinish committed:
Add dynamic scheduling operation mode

This commit adds an opt-in scheduler option for dynamic scheduling.
Instead of partitioning the test list up-front based on historical
timing data this commit lets each worker ask for the next test
dynamically.  This is built using python's multiprocess module to
launch new workers instead of shelling out to call python via
subprocess.

This hopefully will provide a better worker balance since we will keep
each worker occupied until there are no more tests to be run. Instead
of trying to pack fill each work optimially up front. Additionally this
should hopefully improve the pdb story for users who use pdb with tests.
Since instead of spawning subprocesses calling python to invoke the
subunit runner and reading the subunit stream from stdout and instead
uses multiprocessing to fork workers and uses pipes to pass the subunit
streams between workers.

mtreinish commented:
This still fails some tests but after switching away from `subunit.run` this seems to run reliably on python >=3.5. It also will need proper test coverage since this flow is very different from what was there before. I probably should add a dynamic job to travis and appveyor to split out running the full suite between the old way and this new approach.

aspiers commented on stestr/commands/run.py:
```diff
@@ -363,6 +366,13 @@ def run_command(config='.stestr.conf', repo_type='file',
         stdout.write(msg)
         return 2
+    if dynamic:
+        if sys.version_info()[0] < 3 or sys.version_info[1] < 5:
```
For obvious reasons this causes:

    TypeError: 'sys.version_info' object is not callable

```suggestion
        if sys.version_info[0] < 3 or sys.version_info[1] < 5:
```

mtreinish replied:
Oops, good catch thanks

mtreinish committed:
Update stestr/commands/run.py

Co-Authored-By: Adam Spiers <github@adamspiers.org>

aspiers reviewed (commented):
Thanks for accepting my suggestion.  As per [this comment](https://github.com/mtreinish/stestr/issues/14#issuecomment-529433109) this doesn't seem to work for me yet.

mtreinish committed:
Fix failing tests

This commit fixes the failing tests by catching a couple of missing
things from the update. The biggest fix was that for the --no-discover
case we still use a subprocess and because of that we need to tell
output.ReturnCodeToSubunit to that the input is not dynamic (and
therefore a Popen object) so it can handle that properly. The other
major change is that the return code tests are updated so that the
stdout and stderr from the subprocess calls are always decoded in the
non-subunit test cases. This was done primarily for ease of debugging,
but it also enabled the removal of several decode() calls when the
output is parsed.

mtreinish committed:
Clean up test_return_code tests slightly

This is a refinement on the previous commit to reduce unecessary changes
to the functional tests in the test_return_codes module. Mainly always
decoding the output from the subprocess for testing broken things
unexpectedly when a bytes object was expected.

mtreinish committed:
Merge remote-tracking branch 'origin/master' into dynamic-schedule

Conflicts:
	stestr/commands/run.py
	stestr/output.py
	stestr/test_processor.py
	stestr/tests/test_return_codes.py

mtreinish committed:
Merge branch 'master' into dynamic-schedule

mtreinish committed:
Try running with --dynamic everywhere

mtreinish committed:
Merge remote-tracking branch 'origin/main' into dynamic-schedule

mtreinish committed:
Fix pep8

mtreinish committed:
Merge branch 'main' into dynamic-schedule

mtreinish committed:
Merge remote-tracking branch 'origin/main' into dynamic-schedule

mtreinish committed:
Remove check for Python < 3.5

I originally developed this feature when we still supported older
python versions in stestr. The dynamic scheduling feature depends on
functionality added in Python 3.5. Since then the WIP feature branch sat
stale for years since that time we've bumped the minimum version of
Python supported to 3.7 so the runtime check for older python versions
is no longer needed.

mtreinish committed:
Merge branch 'main' into dynamic-schedule

mtreinish committed:
Merge branch 'main' into dynamic-schedule

mtreinish committed:
Merge branch 'main' into dynamic-schedule

mtreinish committed:
Explicitly use spawn for worker processes and fix queue lifetime

This commit fixes an issue that occured in earlier commits on the PR
around the initialization of the worker processes and the scope of the
launch method. Previously if the method used to launch threads returned
before all the workers accessed the queue for the first time the worker
wouldn't be able to read from the queue. This race condition was caused
because the Queue was locally scoped to the method and would be deleted
by the main process before other workers could read it. This would
specifically occur on systems using "forkserver" or "spawn"
multiprocessing start methods because the child processes didn't have
the queue object, while "fork" would because the process memory was
copied in the child process. This commit fixes this by scoping the Queue
object to the instance which means it survives as long as the test
processor object does (which is typically the entire run command).

As part of this change the start method used by the new dynamic
scheduler is set to be fixed to "spawn" to minimize any potential
interactions between stestr and the code under test. This mirrors the
behavior of running in non-dynamic scheduler mode, because spawn is
roughly equivalent to calling python in a subprocess.

mtreinish committed:
Better document the new feature and disable windows testing

This commit improves the documentation of the new --dynamic flag to
explain how it operates and what the goal of it is. It also makes it
clear the feature is experimental and is an opt-in at your own risk.
Also from testing this doesn't currently work on Windows, instead of
blocking the feature over a platform used by 2-3% of our users
(according to https://pypistats.org/packages/stestr ) this just marks it
as currently unsupported. We will have to revisit how to make this work
on Windows before we stabilize the feature.

mtreinish committed:
Fix docs and lint

thomasgoirand commented:
Under Python 3.13, I get:

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/stestr/output.py", line 183, in __del__
    self.proc.join()
AttributeError: 'dict' object has no attribute 'join'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
    from multiprocessing.spawn import spawn_main; spawn_main(tracker_fd=6, pipe_handle=103)
                                                  ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.13/multiprocessing/spawn.py", line 122, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/usr/lib/python3.13/multiprocessing/spawn.py", line 132, in _main
    self = reduction.pickle.load(from_parent)
  File "/usr/lib/python3.13/multiprocessing/synchronize.py", line 115, in __setstate__
    self._semlock = _multiprocessing.SemLock._rebuild(*state)
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory
```

mtreinish committed:
Merge branch 'main' into dynamic-schedule
